File: Catalog REST/src/main/java/cz/vhromada/catalog/rest/controller/SongController.java
package cz.vhromada.catalog.rest.controller;

import cz.vhromada.catalog.facade.SongFacade;
import cz.vhromada.catalog.facade.to.MusicTO;
import cz.vhromada.catalog.facade.to.SongTO;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;

/**
 * A class represents controller for songs.
 *
 * @author Vladimir Hromada
 */
@Controller("songController")
@RequestMapping("/music/{musicId}/songs")
public class SongController extends JsonController {

    @Autowired
    @Qualifier("songFacade")
    private SongFacade songFacade;

    /**
     * Returns song with ID or null if there isn't such song.
     *
     * @param musicId music ID
     * @param songId  song ID
     * @return song with ID or null if there isn't such song
     * @throws IllegalArgumentException if ID is null
     */
    @RequestMapping(value = "/{songId}", method = RequestMethod.GET)
    @ResponseBody
    public String getSong(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, @PathVariable("songId") final Integer songId) {
        return serialize(songFacade.getSong(songId));
    }

    /**
     * Adds song. Sets new ID and position.
     *
     * @param musicId music ID
     * @param song    song
     * @throws IllegalArgumentException                                  if song is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID isn't null
     *                                                                   or name is null
     *                                                                   or name is empty string
     *                                                                   or length of song is negative value
     *                                                                   or note is null
     *                                                                   or music is null
     *                                                                   or music ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if music doesn't exist in data storage
     */
    @RequestMapping(value = "/add", method = RequestMethod.POST)
    @ResponseBody
    public void add(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, final String song) {
        songFacade.add(deserialize(song, SongTO.class));
    }

    /**
     * Updates song.
     *
     * @param musicId music ID
     * @param song    new value of song
     * @throws IllegalArgumentException                                  if song is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     *                                                                   or name is null
     *                                                                   or name is empty string
     *                                                                   or length of song is negative value
     *                                                                   or note is null
     *                                                                   or music is null
     *                                                                   or music ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if song doesn't exist in data storage
     *                                                                   or music doesn't exist in data storage
     */
    @RequestMapping(value = "/update", method = RequestMethod.POST)
    @ResponseBody
    public void update(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, final String song) {
        songFacade.update(deserialize(song, SongTO.class));
    }

    /**
     * Removes song.
     *
     * @param musicId music ID
     * @param song    song
     * @throws IllegalArgumentException                                  if song is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if song doesn't exist in data storage
     */
    @RequestMapping(value = "/remove", method = RequestMethod.POST)
    @ResponseBody
    public void remove(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, final String song) {
        songFacade.remove(deserialize(song, SongTO.class));
    }

    /**
     * Duplicates song.
     *
     * @param musicId music ID
     * @param song    song
     * @throws IllegalArgumentException                                  if song is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if song doesn't exist in data storage
     */
    @RequestMapping(value = "/duplicate", method = RequestMethod.POST)
    @ResponseBody
    public void duplicate(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, final String song) {
        songFacade.duplicate(deserialize(song, SongTO.class));
    }

    /**
     * Moves song in list one position up.
     *
     * @param musicId music ID
     * @param song    song
     * @throws IllegalArgumentException                                  if song is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     *                                                                   or song can't be moved up
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if song doesn't exist in data storage
     */
    @RequestMapping(value = "/moveUp", method = RequestMethod.POST)
    @ResponseBody
    public void moveUp(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, final String song) {
        songFacade.moveUp(deserialize(song, SongTO.class));
    }

    /**
     * Moves song in list one position down.
     *
     * @param musicId music ID
     * @param song    song
     * @throws IllegalArgumentException                                  if song is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     *                                                                   or song can't be moved down
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if song doesn't exist in data storage
     */
    @RequestMapping(value = "/moveDown", method = RequestMethod.POST)
    @ResponseBody
    public void moveDown(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, final String song) {
        songFacade.moveDown(deserialize(song, SongTO.class));
    }

    /**
     * Returns true if song exists.
     *
     * @param musicId music ID
     * @param song    song
     * @return true if song exists
     * @throws IllegalArgumentException                              if song is null
     * @throws cz.vhromada.validators.exceptions.ValidationException if ID is null
     */
    @RequestMapping(value = "/exists", method = RequestMethod.GET)
    @ResponseBody
    public String exists(@PathVariable("musicId") @SuppressWarnings("unused") final Integer musicId, final String song) {
        return serialize(songFacade.exists(deserialize(song, SongTO.class)));
    }

    /**
     * Returns list of songs for specified music.
     *
     * @param musicId music ID
     * @return list of songs for specified music
     * @throws IllegalArgumentException                                  if music is null
     * @throws cz.vhromada.validators.exceptions.ValidationException     if ID is null
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException if music doesn't exist in data storage
     */
    @RequestMapping(value = { "", "/", "list" }, method = RequestMethod.GET)
    @ResponseBody
    public String findSongsByMusic(@PathVariable("musicId") final Integer musicId) {
        final MusicTO music = new MusicTO();
        music.setId(musicId);

        return serialize(songFacade.findSongsByMusic(music));
    }

}

File: catalog-rest/src/main/kotlin/com/github/vhromada/catalog/rest/controller/SongController.kt
package com.github.vhromada.catalog.rest.controller

import com.github.vhromada.catalog.entity.Song
import com.github.vhromada.catalog.facade.MusicFacade
import com.github.vhromada.catalog.facade.SongFacade
import com.github.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for songs.
 *
 * @author Vladimir Hromada
 */
@RestController("songController")
@RequestMapping("/catalog/music/{musicId}/songs")
class SongController(
    private val songFacade: SongFacade,
    private val musicFacade: MusicFacade
) : AbstractController() {

    /**
     * Returns song with ID.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *  * Song doesn't exist in data storage
     *
     * @param musicId music ID
     * @param songId  song ID
     * @return song with ID
     */
    @GetMapping("/{songId}")
    fun getSong(
        @PathVariable("musicId") musicId: Int,
        @PathVariable("songId") songId: Int
    ): Song {
        processResult(result = musicFacade.get(id = musicId))
        return processResult(result = songFacade.get(id = songId))!!
    }

    /**
     * Adds song. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *  * Song ID isn't null
     *  * Song position isn't null
     *  * Name is null
     *  * Name is empty string
     *  * Length of song is negative value
     *  * Note is null
     *
     * @param musicId music ID
     * @param song    song
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(
        @PathVariable("musicId") musicId: Int,
        @RequestBody song: Song
    ) {
        processResult(result = songFacade.add(parent = musicId, data = song))
    }

    /**
     * Updates song.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *  * ID isn't null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * Length of song is negative value
     *  * Note is null
     *  * Song doesn't exist in data storage
     *
     * @param musicId music ID
     * @param song    new value of song
     */
    @PostMapping("/update")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun update(
        @PathVariable("musicId") musicId: Int,
        @RequestBody song: Song
    ) {
        processResult(result = musicFacade.get(id = musicId))
        processResult(result = songFacade.update(data = song))
    }

    /**
     * Removes song.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *  * Song doesn't exist in data storage
     *
     * @param musicId music ID
     * @param id      ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(
        @PathVariable("musicId") musicId: Int,
        @PathVariable("id") id: Int
    ) {
        processResult(result = musicFacade.get(id = musicId))
        processResult(result = songFacade.remove(id = id))
    }

    /**
     * Duplicates song.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *  * Song doesn't exist in data storage
     *
     * @param musicId music ID
     * @param id      ID
     */
    @PostMapping("/duplicate/{id}")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(
        @PathVariable("musicId") musicId: Int,
        @PathVariable("id") id: Int
    ) {
        processResult(result = musicFacade.get(id = musicId))
        processResult(result = songFacade.duplicate(id = id))
    }

    /**
     * Moves song in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *  * Song can't be moved up
     *  * Song doesn't exist in data storage
     *
     * @param musicId music ID
     * @param id      ID
     */
    @PostMapping("/moveUp/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(
        @PathVariable("musicId") musicId: Int,
        @PathVariable("id") id: Int
    ) {
        processResult(result = musicFacade.get(id = musicId))
        processResult(result = songFacade.moveUp(id = id))
    }

    /**
     * Moves song in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *  * Song can't be moved down
     *  * Song doesn't exist in data storage
     *
     * @param musicId music ID
     * @param id      ID
     */
    @PostMapping("/moveDown/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(
        @PathVariable("musicId") musicId: Int,
        @PathVariable("id") id: Int
    ) {
        processResult(result = musicFacade.get(id = musicId))
        processResult(result = songFacade.moveDown(id = id))
    }

    /**
     * Returns list of songs for specified music.
     * <br></br>
     * Validation errors:
     *
     *  * Music doesn't exist in data storage
     *
     * @param musicId music ID
     * @return list of songs for specified music
     */
    @GetMapping
    fun findSongsByMusic(@PathVariable("musicId") musicId: Int): List<Song> {
        return processResult(result = songFacade.find(parent = musicId))!!
    }

}
